src/Hosting/Hosting/src/StaticWebAssets/StaticWebAssetsFileProvider.cs

```diff
@@ -158,4 +157,362 @@ public Stream CreateReadStream()
+                    candidate = null;
+                    break;
+                }
+            }
+
+            var match = candidate?.Match;
+            if (match != null)
+            {
+                // If we found a file, that wins over anything else. If there are conflicts with files added after
+                // we've built the manifest, we'll be notified the next time we do a build. This is not different
+                // from previous Static Web Assets versions.
+                var file = _fileProviders[match.ContentRoot].GetFileInfo(match.Path);
+                if (!file.Exists || string.Equals(subpath, Normalize(match.Path), _fsComparison))
+                {
+                    return file;
+                }
+                else
+                {
+                    return new StaticWebAssetsFileInfo(segments[^1], file);
+                }
+            }
+
+            // The list of patterns is ordered by pattern depth, so we compute the string to check for patterns only
+            // once per level. We don't aim to solve conflicts here where multiple files could match a given path,
+            // we have a build check that takes care of that.
+            var currentDepth = 0;
+            var candidatePath = subpath;
+
+            if (patterns != null)
+            {
+                for (var i = 0; i < patterns.Count; i++)
+                {
+                    var pattern = patterns[i];
+                    if (pattern.Depth != currentDepth)
+                    {
+                        currentDepth = pattern.Depth;
+                        candidatePath = string.Join('/', segments[currentDepth..]);
+                    }
+
+                    var result = _fileProviders[pattern.ContentRoot].GetFileInfo(candidatePath);
+                    if (result.Exists)
+                    {
+                        if (!result.IsDirectory)
+                        {
+                            var matcher = new Matcher();
+                            matcher.AddInclude(pattern.Pattern);
+                            if (!matcher.Match(candidatePath).HasMatches)
+                            {
+                                continue;
+                            }
+                        }
+
+                        return result;
+                    }
+                }
+            }
+
+            if (candidate != null)
+            {
+                // we didn't find a match, nor a match on the patterns, but at least we matched the segments against the tree
+                // on the manifest, so this indicates it is a folder.
+                return new StaticWebAssetsDirectoryInfo(segments.Length == 0 ? "" : segments[^1]);
+            }
+
+            return new NotFoundFileInfo(subpath);
+        }
+
+        public IChangeToken Watch(string filter) => NullChangeToken.Singleton;
+
+        private record StaticWebAssetsDirectoryContents(IEnumerable<IFileInfo> Files) : IDirectoryContents
```

Comment:
Is this a record because you rely on the equality? If not, it would be preferable to use POCO types here. The added IL isn't really worth the trouble to allow using primary ctors.

Reply:
Just because it was less code.

I don't think it matters on the server, however I'll change it